File: src/17-challenge/solution.py
from typing import Literal
import re

Light = Literal[".", "R", "G"]


def fn(board: list[list[Light]]) -> bool:
    for horizontal_line in board:
        has_four_horizontal_lights = re.search("R{4}|G{4}", "".join(horizontal_line))
        if has_four_horizontal_lights:
            return True

    for i in range(len(board[0])):
        vertical_line = board[0][i]
        for j in range(1, len(board)):
            vertical_line += board[j][i]

        has_four_vertical_lights = re.search("R{4}|G{4}", vertical_line)
        if has_four_vertical_lights:
            return True

    return False

File: src/17-challenge/solution.js
/**
 * @typedef {'.' | 'R' | 'G'} Light
 */

/**
 * @param {Light[][]} board
 * @returns {boolean}
 */
export default function fn(board) {
	for (const horizontalLine of board) {
		const hasFourHorizontalLights = horizontalLine.join('').match(/R{4}|G{4}/)
		if (hasFourHorizontalLights) return true
	}

	for (let i = 0; i < (board[0] ?? []).length; i++) {
		let verticalLine = board[0][i]
		for (let j = 1; j < board.length; j++) verticalLine += board[j][i]

		const hasFourVerticalLights = verticalLine.match(/R{4}|G{4}/)
		if (hasFourVerticalLights) return true
	}

	return false
}
